File: go/0019_remove_nth_node_from_end_of_list.go
package main

func removeNthFromEnd(head *ListNode, n int) *ListNode {
	fast, slow := head, head
	for i := 0; i < n; i++ {
		if fast.Next == nil {
			// assuming all n is valid, so this case
			// the first element to be deleted
			return head.Next
		}
		fast = fast.Next
	}

	for fast.Next != nil {
		fast, slow = fast.Next, slow.Next
	}
	slow.Next = slow.Next.Next
	return head
}

File: python/0019_remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """
        
        if not head:
            return None
        
        length = 0
        cur = head
        while cur:
            length += 1
            cur = cur.next
        
        target = length - n
        i = 0
        prev, cur = None, head
        while i < target:
            prev = cur
            cur = cur.next
            i += 1
        
        if cur == head:
            head = cur.next
        else:
            prev.next = cur.next
        return head